### PR TITLE
Add Crown Jewel resource

### PR DESCRIPTION
**Pull request by jakule**
Add Crown Jewel resource described here https://github.com/gravitational/access-graph/blob/main/rfd/0005-crown-jewel.md

Depends on https://github.com/gravitational/teleport/pull/41168

changelog: Add support for crown jewel resource type

### CODE REVIEW

File: api/types/crownjewel/crownjewel.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/api/types/header"
+	"github.com/gravitational/teleport/api/types/header/convert/legacy"
+	"github.com/gravitational/teleport/api/utils"
+)
+
+var _ types.Resource = &CrownJewel{}
+
+// CrownJewel is a resource that represents the crown jewel resource.
+type CrownJewel struct {
```
[the remark this answers is not in the document]
Comment (jakule) [reply api/types/crownjewel/crownjewel.go:13]: Sure, I can update it. But our implementation are really getting messy as I copied the pattern from access list and discovery config that AFAIU were a "popular pattern" last year.

File: lib/services/local/crown_jewels_test.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+	}
+
+	out, err := MarshalCrownJewel(obj)
+	require.NoError(t, err)
+	newObj, err := UnmarshalCrownJewel(out)
+	require.NoError(t, err)
+	require.True(t, proto.Equal(obj, newObj), "messages are not equal")
+}
+
+//TODO(jaku): add more tests
```
[the remark this answers is not in the document]
Comment (jakule) [reply lib/services/local/crown_jewels_test.go:49]: My bad, I added this test in a different branch by committed here 🤦

File: tool/tctl/common/resource_command.go [hunk 3]
```diff
@@ -911,6 +913,28 @@ func (rc *ResourceCommand) createKubeCluster(ctx context.Context, client *auth.C
...
+				return trace.AlreadyExists("crown jewel %q already exists", crownJewel.GetMetadata().GetName())
+			}
+			if _, err := client.CrownJewelsClient().UpdateCrownJewel(ctx, crownJewel); err != nil {
+				return trace.Wrap(err)
+			}
+			fmt.Printf("crown jewel %q has been updated\n", crownJewel.GetMetadata().GetName())
+			return nil
+		}
+		return trace.Wrap(err)
+	}
```
[the remark this answers is not in the document]
Comment (jakule) [reply tool/tctl/common/resource_command.go:933]: It worked when I tested it and I also call update here when create fails.

File: lib/auth/crownjewel/crownjewelv1/service.go [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+	authCtx, err := s.authorizer.Authorize(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if err := authCtx.CheckAccessToKind(types.KindCrownJewel, types.VerbRead, types.VerbList); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	rsp, nextToken, err := s.backend.ListCrownJewels(ctx, req.PageSize, req.PageToken)
```
[the remark this answers is not in the document]
Comment (jakule) [reply lib/auth/crownjewel/crownjewelv1/service.go]: Changed, I also updated the `Get` method to read from cache

File: lib/services/local/crown_jewels.go [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+	return r, nextToken, trace.Wrap(err)
+}
+
+func (s *CrownJewelsService) CreateCrownJewel(ctx context.Context, crownJewel *crownjewelv1.CrownJewel) (*crownjewelv1.CrownJewel, error) {
+	r, err := s.service.CreateResource(ctx, crownJewel)
+	return r, trace.Wrap(err)
+}
+
+func (s *CrownJewelsService) UpdateCrownJewel(ctx context.Context, crownJewel *crownjewelv1.CrownJewel) (*crownjewelv1.CrownJewel, error) {
+	r, err := s.service.UpdateResource(ctx, crownJewel)
```
[the remark this answers is not in the document]
Comment (jakule) [reply lib/services/local/crown_jewels.go:68]: I'll update it after your PR is merged.

**Review by rosstimothy — APPROVED** [review]

File: lib/services/local/crown_jewels.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (jakule) [reply lib/services/local/crown_jewels.go:68]: Updated.

**Review by tigrato — APPROVED** [review]

File: lib/auth/crownjewel/crownjewelv1/service_test.go [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+	for state := range allAdminStates {
+		found := slices.Index(states, state) != -1
+		if !found {
+			out = append(out, state)
+		}
+	}
+	return out
+}
+
+// callMethod calls a method with given name in the DatabaseObjectService service
```
Comment (espadolini) [inline lib/auth/crownjewel/crownjewelv1/service_test.go:164]:
```suggestion
// callMethod calls a method with given name in the CrownJewelService service.
```

File: lib/auth/crownjewel/crownjewelv1/service_test.go [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+func callMethod(t *testing.T, service *Service, method string) error {
```
Comment (espadolini) [inline lib/auth/crownjewel/crownjewelv1/service_test.go:165]:
```suggestion
func callMethod(t *testing.T, service crownjewelv1.CrownJewelServiceServer, method string) error {
```

File: lib/auth/crownjewel/crownjewelv1/service_test.go [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+	for _, desc := range crownjewelv1.CrownJewelService_ServiceDesc.Methods {
```
Comment (espadolini) [inline lib/auth/crownjewel/crownjewelv1/service_test.go:166]: The doc for `ServiceName_ServiceDesc` states

```
// It's only intended for direct use with grpc.RegisterService,
// and not to be introspected or modified (even as a copy)
```

but I can't find a useful way to make use of `protoreflect`/`protoregistry`, and I'm not entirely sure if going through `reflect` is better than going through an exported var that Must Not Be Used (and that other projects are indubitably also using).

<details>
<summary>for posterity</summary>

```go
	m := reflect.ValueOf(service).MethodByName(method)
	if !m.IsValid() {
		require.FailNow(t, "method %v not found", method)
	}
	r := m.Call([]reflect.Value{
		reflect.ValueOf(context.Background()),
		reflect.New(m.Type().In(1).Elem()),
	})
	err := r[1].Interface()
	if err == nil {
		return nil
	}
	return err.(error)
```

</details>

File: lib/auth/crownjewel/crownjewelv1/service_test.go [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+	services.AccessChecker
+}
+
+func (f fakeChecker) CheckAccessToRule(_ services.RuleContext, _ string, resource string, verb string) error {
+	if resource == types.KindCrownJewel {
+		for _, allowedVerb := range f.allowedVerbs {
+			if allowedVerb == verb {
+				return nil
+			}
+		}
```
Comment (espadolini) [inline lib/auth/crownjewel/crownjewelv1/service_test.go:187]:
```suggestion
		if slices.Contains(f.allowedVerbs, verb) {
			return nil
		}
```

File: lib/auth/crownjewel/object.go [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+	}
+
+	if err := ValidateCrownJewel(cj); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return cj, nil
+}
+
+// ValidateCrownJewel validates the CrownJewel object.
```
Comment (espadolini) [inline lib/auth/crownjewel/object.go:45]: I think it's useful if we start documenting this sort of details.

```suggestion
// ValidateCrownJewel validates the CrownJewel object without modifying.
```

File: lib/auth/crownjewel/object.go [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+// Required fields:
+//   - Metadata.Name
+//   - Spec.TeleportMatchers or Spec.AwsMatchers
+//   - Matcher.Name or Matcher.Labels
+func ValidateCrownJewel(jewel *crownjewelv1.CrownJewel) error {
+	switch {
+	case jewel == nil:
```
Comment (espadolini) [inline lib/auth/crownjewel/object.go:52]: for this and the later parts: a `nil` pointer protobuf message is a valid read-only message that has all fields zeroed; should we start using getters whenever possible? we've kind of avoided that for gogoproto, which doesn't always generate getters and can store submessages as structs, but that's not how the convention goes for protobuf

File: lib/auth/crownjewel/object.go [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+			if len(matcher.GetKinds()) == 0 {
+				return trace.BadParameter("teleport matcher kinds must be set")
+			}
+
+			if matcher.Name == "" && len(matcher.GetLabels()) == 0 {
+				return trace.BadParameter("teleport matcher name or labels must be set")
+			}
+
+			if len(matcher.GetLabels()) > 0 {
+				for _, label := range matcher.GetLabels() {
```
Comment (espadolini) [inline lib/auth/crownjewel/object.go:75]: Unnecessary check

File: lib/auth/crownjewel/object.go [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+			if len(matcher.GetTypes()) == 0 {
+				return trace.BadParameter("aws matcher type must be set")
+			}
+
+			if matcher.GetArn() == "" && len(matcher.GetTags()) == 0 {
+				return trace.BadParameter("aws matcher arn or tags must be set")
+			}
+
+			if len(matcher.GetTypes()) > 0 {
+				for _, label := range matcher.GetTags() {
```
Comment (espadolini) [inline lib/auth/crownjewel/object.go:95]: Typo, but even if the check was for GetTags it'd be unnecessary 🤣

File: lib/cache/cache.go [hunk 14]
```diff
@@ -2811,7 +2823,33 @@ func (c *Cache) GetDiscoveryConfig(ctx context.Context, name string) (*discovery
...
+func (c *Cache) GetCrownJewel(ctx context.Context, name string) (*crownjewelv1.CrownJewel, error) {
+	ctx, span := c.Tracer.Start(ctx, "cache/GetCrownJewel")
+	defer span.End()
+
+	rg, err := readCollectionCache(c, c.collections.crownJewels)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer rg.Release()
+	return rg.reader.GetCrownJewel(ctx, name)
```
Comment (espadolini) [inline lib/cache/cache.go:2849]: Would it make sense to read through the cache on a `NotFound`? We do that for some resources like roles, for example.

File: lib/services/crown_jewel.go [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+func MarshalCrownJewel(object *crownjewelv1.CrownJewel, opts ...MarshalOption) ([]byte, error) {
+	cfg, err := CollectOptions(opts)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if !cfg.PreserveResourceID {
+		object = proto.Clone(object).(*crownjewelv1.CrownJewel)
+		object.Metadata.Revision = ""
+	}
+	data, err := utils.FastMarshal(object)
```
Comment (espadolini) [inline lib/services/crown_jewel.go:60]: no protojson? 🥺

**Review by espadolini — APPROVED** [review]

File: lib/auth/crownjewel/crownjewelv1/service_test.go (hunk 7 above)
Comment (jakule) [reply lib/auth/crownjewel/crownjewelv1/service_test.go:165]: This breaks compilation and I don't want to update all types now, as we do that in all other services.

File: lib/auth/crownjewel/crownjewelv1/service_test.go (hunk 8 above)
Comment (jakule) [reply lib/auth/crownjewel/crownjewelv1/service_test.go:166]: >and not to be introspected or modified

Reflect sounds like introspecting 🤔 I'll leave the current version as it looks "safer" IMO

**Review by r0mant — APPROVED** [review]
Bot.

**Pull request merged**